packages/@aws-cdk/region-settings/lib/service-principal.ts

```diff
@@ -0,0 +1,50 @@
+import { Region } from "./region";
+
+/**
+ * An entity that is capable of computing service principals (e.g: sqs.amazonaws.com).
+ */
+export interface IServicePrincipalMaker {
+  /**
+   * Coputes the appropriate service principal for a given service in a given region.
+   * @param service the short name of the service for which a principal is needed (e.g: `sqs`)
+   * @param region  the region in where the service principal resides
+   * @returns a service principal string (e.g: `sqs.amazonaws.com`)
+   */
+  makeServicePrincipal(service: string, region: Region): string;
+}
+
+/**
+ * A service principal maker that knows to generate valid principals for the built-in AWS region settings.
+ */
+export class DefaultServicePrincipalMaker implements IServicePrincipalMaker {
+  public static instance: IServicePrincipalMaker = new DefaultServicePrincipalMaker();
+
+  private constructor() { }
+
+  public makeServicePrincipal(service: string, region: Region): string {
```

Comment:
None of these template strings are true in nonpublic regions -- all the principals there are nontemplatable, one-off strings.

How will the current model cope with those?

Reply:
I think the only way to model this information in an extensible and manageable way is going to be using a rules engine.

Reply:
Examples please? Because what I've seen so far suggests otherwise.